packages/opencensus-instrumentation-http2/src/http2.ts

```diff
@@ -0,0 +1,283 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {HttpPlugin} from '@opencensus/instrumentation-http';
+import {types} from '@opencensus/opencensus-core';
+import {classes} from '@opencensus/opencensus-core';
+import {logger} from '@opencensus/opencensus-core';
+import * as http from 'http';
+import * as http2 from 'http2';
+import * as net from 'net';
+import * as shimmer from 'shimmer';
+import * as tls from 'tls';
+import * as url from 'url';
+import * as uuid from 'uuid';
+
+export type Http2Module = typeof http2;
+export type ConnectFunction = typeof http2.connect;
+type RequestFunction =
+    (this: http2.ClientHttp2Session, headers?: http2.OutgoingHttpHeaders,
+     options?: http2.ClientSessionRequestOptions) => http2.ClientHttp2Stream;
+export type CreateServerFunction = typeof http2.createServer;
+
+/** Http2 instrumentation plugin for Opencensus */
+export class Http2Plugin extends HttpPlugin {
+  server: http2.Http2Server|http2.Http2SecureServer;
+  client: http2.ClientHttp2Session;
+  clientUrl: url.UrlWithStringQuery;
+  patchedCreateServer = false;
+  patchedConnect = false;
+
+  /** Constructs a new Http2Plugin instance. */
+  constructor() {
+    super('http2');
+  }
+
+  /**
+   * Patches HTTP2 incoming and outcoming request functions.
+   * @param moduleExporters The HTTP2 package.
+   * @param tracer A tracer instance to create spans on.
+   * @param version The package version.
+   */
+  // tslint:disable-next-line:no-any
+  applyPatch(moduleExports: any, tracer: types.Tracer, version: string) {
+    this.setPluginContext(moduleExports, tracer, version);
+    this.logger = tracer.logger || logger.logger('debug');
+
+    shimmer.wrap(moduleExports, 'createServer', this.patchCreateServer());
+    shimmer.wrap(moduleExports, 'createSecureServer', this.patchCreateServer());
+
+    shimmer.wrap(moduleExports, 'connect', this.patchConnect());
+
+    return moduleExports;
+  }
+
+  /** Unpatches all HTTP2 patched function. */
+  applyUnpatch(): void {
+    if (this.patchedCreateServer) {
+      shimmer.unwrap(plugin.server.constructor.prototype, 'emit');
+    } else {
+      shimmer.unwrap(plugin.moduleExports, 'createServer');
+      shimmer.unwrap(plugin.moduleExports, 'createSecureServer');
+    }
+
+    if (this.patchedConnect) {
+      shimmer.unwrap(plugin.client, 'request');
+    } else {
+      shimmer.unwrap(plugin.moduleExports, 'connect');
+    }
+  }
+
+  patchConnect() {
+    const plugin = this;
+    return (original: ConnectFunction):
+               types.Func<http2.ClientHttp2Session> => {
+      return function patchedConnect(authority: string):
+          http2.ClientHttp2Session {
+            plugin.client = original.apply(this, arguments);
+            shimmer.wrap(plugin.client, 'request', plugin.patchRequest());
```

Comment:
Is `request` the only way to communicate with an HTTP2 server? I believe there is a low-level API as well.

Reply:
Hmm I couldn't find it. Please let me know if there is, cause we should patch it too.

Reply:
Never mind -- I was mistaken.